go-controller/pkg/metrics/master.go

```diff
@@ -232,147 +245,149 @@ var metricPortBindingUpLatency = prometheus.NewHistogram(prometheus.HistogramOpt
+	prometheus.MustRegister(metricV4AllocatedHostSubnetCount)
+	prometheus.MustRegister(metricV6AllocatedHostSubnetCount)
+	prometheus.MustRegister(metricEgressIPCount)
+	prometheus.MustRegister(metricEgressFirewallRuleCount)
+	prometheus.MustRegister(metricEgressFirewallCount)
+	prometheus.MustRegister(metricEgressRoutingViaHost)
 }
 
-// StartMasterMetricUpdater adds a goroutine that updates a "timestamp" value in the
-// nbdb every 30 seconds. This is so we can determine freshness of the database.
-// Also, update IPsec enabled or disable metric.
-func StartMasterMetricUpdater(stopChan <-chan struct{}, nbClient libovsdbclient.Client) {
-	startMasterMetricUpdaterOnce.Do(func() {
-		addIPSecMetricHandler(nbClient)
-		go func() {
-			ticker := time.NewTicker(30 * time.Second)
-			defer ticker.Stop()
-			for {
-				select {
-				case <-ticker.C:
-					updateE2ETimestampMetric(nbClient)
-				case <-stopChan:
-					return
+// RunTimestamp adds a goroutine that registers and updates timestamp metrics.
+// This is so we can determine 'freshness' of the components NB/SB DB and northd.
+// Function must be called once.
+func RunTimestamp(stopChan <-chan struct{}, sbClient, nbClient libovsdbclient.Client) {
+	// Metric named nb_e2e_timestamp is the UNIX timestamp this instance wrote to NB DB. Updated every 30s with the
+	// current timestamp.
+	prometheus.MustRegister(metricNbE2eTimestamp)
+
+	// Metric named sb_e2e_timestamp is the UNIX timestamp observed in SB DB. The value is read from the SB DB
+	// cache when metrics HTTP endpoint is scraped.
+	scrapeOvnSbE2eTimestamp := func() float64 {
+		sbGlobal, err := libovsdbops.FindSBGlobal(sbClient)
+		if err != nil {
+			klog.Errorf("Failed to get global options for the SB_Global table: %v", err)
+			return 0
+		}
+		if val, ok := sbGlobal.Options["e2e_timestamp"]; ok {
+			return parseMetricToFloat(MetricOvnkubeSubsystemMaster, "sb_e2e_timestamp", val)
+		}
+		return 0
+	}
+	prometheus.MustRegister(prometheus.NewGaugeFunc(
+		prometheus.GaugeOpts{
+			Namespace: MetricOvnkubeNamespace,
+			Subsystem: MetricOvnkubeSubsystemMaster,
+			Name:      "sb_e2e_timestamp",
+			Help:      "The current e2e-timestamp value as observed in the southbound database",
+		}, scrapeOvnSbE2eTimestamp))
+
+	// Metric named e2e_timestamp is the UNIX timestamp observed in NB and SB DBs cache with the DB name
+	// (OVN_Northbound|OVN_Southbound) set as a label. Updated every 30s.
+	prometheus.MustRegister(metricDbTimestamp)
+
+	go func() {
+		ticker := time.NewTicker(30 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				currentTime := time.Now().Unix()
+				if setNbE2eTimestamp(nbClient, currentTime) {
+					metricNbE2eTimestamp.Set(float64(currentTime))
 				}
+
+				metricDbTimestamp.WithLabelValues(nbClient.Schema().Name).Set(getDbOptionsTimestamp(nbClient))
+				metricDbTimestamp.WithLabelValues(sbClient.Schema().Name).Set(getDbOptionsTimestamp(sbClient))
```

Comment:
Due to my ignorance and that I could not find ovn doc on `e2e_timestamp` field, how does this work. 
We set a timestamp in NB and we record it. And then we read them from NB and SB. Is this to measure latency in between?

Reply:
Yes, Its confusing the way its put together due to 4 different metric _values_ for timestamp.
We first of all have `nb_e2e_timestamp`. This stores the timestamp written to NB DB.
We then have `sb_e2e_timestamp` which pulls whatever SB timestamp we have in the libovsdb cache currently when the metrics endpoint is scraped by whatever agent.

We then have `e2e_timestamp` that contains 2 labels: OVN_Northbound and OVN_Southbound. This contains the values that we see in the two respective OVN DBs.

I don't know why `e2e_timestamp` was implemented but I cannot remove it now because folks may use it.

The goal of all of this originally I believe was to see if OVN NB and OVN SB are r/w-able and detect if northd is functioning and it gives us a rough indication how long it took to translate OVN NB config to SB flows.

All this isnt really needed anymore because we have stopwatch metrics for northd and we have OVN DB cluster status metrics for NB and SB. However, I am not going to just remove it incase people use it.